src/extensions/AssetMigration.sol

```diff
@@ -0,0 +1,250 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.27;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IERC721 } from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import { HasProxyAdmin } from "src/extensions/collections/HasProxyAdmin.sol";
+
+import { IWETH } from "src/interfaces/IWETH.sol";
+import { ErrInvalidStorageLocation, ErrLengthMismatch, ErrEmptyArray } from "src/utils/CommonErrors.sol";
+
+abstract contract AssetMigration is HasProxyAdmin {
+  using SafeERC20 for IERC20;
+
+  /// @dev Error when the caller is not authorized
+  error ErrUnauthorizedCaller(address expected, address caller);
+  /// @dev Error when the token is not whitelisted before
+  error ErrNotWhitelistedToken(address token);
+  /// @dev Error when the native token is whitelisted instead of the wrapped token
+  error ErrWhitelistWrappedTokenInstead();
+
+  /// @dev The native token indicator address
+  address internal constant _NATIVE_TOKEN_INDICATOR = address(0);
+
+  /// @custom:storage-location erc7201:ronin.bridge.AssetMigration
+  struct AssetMigrationStorage {
+    IWETH _wnt;
+    // Migrator address
+    address _addr;
+    // Whitelisted addresses
+    mapping(address token => address whitelist) _whitelist;
+  }
+
+  /// @dev Emitted when a new migrator is set.
+  event NewMigrator(address indexed by, address indexed migrator);
+  /// @dev Emitted when the wrapped native token is set.
+  event WrappedNativeTokenSet(address indexed by, address indexed wnt);
+  /// @dev Emitted when recipients are whitelisted.
+  event WhitelistUpdated(address indexed by, address[] tokens, address[] recipients);
+
+  /**
+   * @dev Modifier to check if the caller is the migrator.
+   */
+  modifier onlyMigrator() {
+    _requireMigrator();
+    _;
+  }
+
+  /**
+   * @dev Modifier to check if `a` and `b` have the same length and are not empty.
+   */
+  modifier validInput(uint256[] memory a, uint256[] memory b) {
+    _requireValidInput(a, b);
+    _;
+  }
+
+  /**
+   * @dev Returns the wrapped native token.
+   * - For `RoninGatewayV3`, it MUST return the WRON token.
+   * - For `MainchainGatewayV3`, it MUST return the WETH token.
+   */
+  function _getWrappedNativeToken() internal view returns (IWETH) {
```

Comment:
I think bridge have nothing to do with WRON, we can simply hardcode the WETH if needed (mainnet/testnet)

Reply:
On Ethereum, we've already had `wrappedNativeToken()`.

Reply:
in anycase will we need to migrate RON out of `RoninGatewayV3`?

Reply:
Our bridge do not support bridging RON